compiler/rustc_typeck/src/collect/type_of.rs

```diff
@@ -134,7 +134,15 @@ pub(super) fn opt_const_param_of(tcx: TyCtxt<'_>, def_id: LocalDefId) -> Option<
             | Node::Pat(_) => {
                 let path = match parent_node {
                     Node::Ty(&Ty { kind: TyKind::Path(QPath::Resolved(_, path)), .. })
-                    | Node::TraitRef(&TraitRef { path, .. }) => &*path,
+                    | Node::TraitRef(&TraitRef { path, .. }) => {
+                        // Check for empty generic args, because if there are no generic args, this would otherwise ICE
+                        if path.segments.len() == 1
```

Comment:
Not sure if ```== 1``` is correct here or if checking for ```< 2``` would be better.

Reply:
Checking for the length of the segment does prevent the `unwrap` triggered ICE in this test case, but this doesn't seem like the best way to prevent similar ICEs. I think something like `0: u8<e::f<5>=e::f>` should also trigger an ICE after your fixes. 

Obviously this fuzzer generated example is really contrived, but to prevent this type of ICE, the most reasonable change to me seems to be to shift the code [here](https://github.com/rust-lang/rust/blob/27d5935df18d7035c9bd8dc5c1dfbbad6ac59793/compiler/rustc_typeck/src/collect/type_of.rs#L175-L187), which tries to find the `arg_index` and `segment` from the path, to this pattern arm and not call `unwrap` on the result, but just return None if this fails to find any `arg_index`. Maybe try to minimize code duplication here. 

I'm not sure about the other pattern arms, maybe calling `unwrap` on the resulting path for those branches could also ICE and maybe we should never call `unwrap` in the first place.

Reply:
First of all thanks for the feedback, i'll look into applying the changes you suggested.
(btw ```0: u8<e::f<5>=e::f>``` doesn't panic at all neither on stable nor with my changed)

Reply:
What about `0: u8<e<5>=e::f>`?

Reply:
Stable panics, with my changes it doesn't.

Reply:
> 



> 

You mentioned "this pattern arm" do you mean the pattern arm at line 137 with that?

Reply:
Yes. 

Maybe I'm misunderstanding something here. What is `parent_node` when you run this with `0: u8<e::f<5>=e::f>`?

Reply:
This is parent_node: ```parent: Ty(Ty { hir_id: HirId { owner: DefId(0:3 ~ anon_const[dfde]::main), local_id: 2 }, kind: Path(Resolved(None, Path { span: anon_const.rs:2:8: 2:21 (#0), res: PrimTy(Uint(U8)), segments: [PathSegment { ident: u8#0, hir_id: Some(HirId { owner: DefId(0:3 ~ anon_const[dfde]::main), local_id: 9 }), res: Some(Err), args: Some(GenericArgs { args: [], bindings: [TypeBinding { hir_id: HirId { owner: DefId(0:3 ~ anon_const[dfde]::main), local_id: 8 }, ident: e#0, gen_args: GenericArgs { args: [Const(ConstArg { value: AnonConst { hir_id: HirId { owner: DefId(0:3 ~ anon_const[dfde]::main), local_id: 3 }, body: BodyId { hir_id: HirId { owner: DefId(0:3 ~ anon_const[dfde]::main), local_id: 4 } } }, span: anon_const.rs:2:13: 2:14 (#0) })], bindings: [], parenthesized: false, span_ext: anon_const.rs:2:12: 2:15 (#0) }, kind: Equality { ty: Ty { hir_id: HirId { owner: DefId(0:3 ~ anon_const[dfde]::main), local_id: 5 }, kind: Path(Resolved(None, Path { span: anon_const.rs:2:16: 2:20 (#0), res: Err, segments: [PathSegment { ident: e#0, hir_id: Some(HirId { owner: DefId(0:3 ~ anon_const[dfde]::main), local_id: 6 }), res: Some(Err), args: None, infer_args: false }, PathSegment { ident: f#0, hir_id: Some(HirId { owner: DefId(0:3 ~ anon_const[dfde]::main), local_id: 7 }), res: Some(Err), args: None, infer_args: false }] })), span: anon_const.rs:2:16: 2:20 (#0) } }, span: anon_const.rs:2:11: 2:20 (#0) }], parenthesized: false, span_ext: anon_const.rs:2:10: 2:21 (#0) }), infer_args: false }] })), span: anon_const.rs:2:8: 2:21 (#0) })```

Reply:
> 

But wouldn't moving that code block to said pattern lead to code duplication and to computing the result of that code block twice? Is there a reason why I can't just let the code block be where it is and replace the ```bug!``` with a ```return None``` and if necessary add a boolean which indicates from which pattern arm the ```path``` came from?

Reply:
> Is there a reason why I can't just let the code block be where it is and replace the bug! with a return None

I would assume that there is a reason for that. Returning None instead of unwrap would certainly be the cleanest solution, but I'm not sure in what contexts `opt_const_param_of` is called and how this call is handled in case we cannot find the generic parameter corresponding to this path and return None. It might be that we don't handle some error correctly if we just return None whenever we cannot find a generic param. Can you look into that?

Reply:
"I would assume that there is a reason for that" I don't really understand what you mean by that, but sure, i'll look into the context thingy.

Reply:
I just mean that without knowing more about this code that it makes more sense to assume that the person who inserted that `unwrap` probably had a good reason for including it.

Reply:
So, i looked into it and i didn't find any case in which it could cause problems
Disclaimer: I am relatively new to rustc development, so maybe i missed something.

Reply:
@jackh726 what are your thoughts on this?

Reply:
Right, so I guess in this case `u8<e<5>=e>` is treated as type ascription for the `0`?

I think the right thing here is to change the `unwrap_or_else(|| bug!())` [here](https://github.com/rust-lang/rust/blob/27d5935df18d7035c9bd8dc5c1dfbbad6ac59793/compiler/rustc_typeck/src/collect/type_of.rs#L186) to a `delay_span_bug` ("no arg matching AnonConst in path") and a `return None`.

Reply:
> 

Right, so i applied your suggestions, I had to add another variable cuz i can't return the ```None``` value in the ```unwrap_or_else```, also the test output didn't change